### ISSUE TITLE
Project team needed?

### ISSUE THREAD

**Issue by JackDca** [mailpile/Mailpile#2223]
In @BjarniRunar's blog post [Burned Out and Happy?](https://www.mailpile.is/blog/2019-04-06_Burnout.html) he says,
 "I care deeply about this project, but it really is too big for one
person and I haven't successfully built a team to support me."

I too have spent some time on Mailpile (but nothing like
Bjarni's contribution!). I believe that it is an important
effort and want to see it continue. Can this community have a
discussion about action we can take to create the team as
suggested by Bjarni?

In the world of engineering management, it is recognized that
the most effective way to get a project done is often to have a
single person do everything from function definition to detailed
design and testing. No time is used up in communicating between
team members, debating interfaces and so on. But this only works
to the point where the project is small enough for one person.
Although Mailpile has had many contributors, Bjarni's share far
exceeds anyone else's so it is not far from being a one person
project, but it has grown past the point where that structure
works.
 
Creating a workable project team is not a trivial task, but it
seems desirable at this point. Here are some ideas. I'm sure
that they are not new, but it would be good to see some
discussion.

1. Most important, will any of us make an ongoing commitment
to write or modify code to fix specific issues? It might help if
issues could be prioritized and contributors recruited for e.g.
the five highest priority issues.

2. Bjarni has so far kept Mailpile going in a consistent
direction by personally reviewing all the code. Could we find
some additional volunteers who are qualified to do some or all
of the code reviews, so that Bjarni only has to do the actual
commits? For this to work, there would need to be some high
level discussion and agreement on the functions being
implemented, and programming standards, to guide the reviewers
and keep everyone consistent. If this works well, perhaps commit
authority could eventually be delegated to some of the
reviewers.

3. Can we find volunteers who would take responsibility for
responding to Github issues and to questions on the Community
site? It would be best if there were a few of these, each with
an area of expertise, so that more specialized questions could
be handed off to an appropriate person.

4. Could we find people in the community who could document some
of the internals of Mailpile? I find that I spend more time
figuring out how the code works than actually modifying it. More
documentation would result in more efficient use of developer
time and speed up code reviews.

One way to attract more developers is to engage more users and
contributors. This should only be done after the project team
has been created. There's no point in attracting more people if
their comments, requests for assistance and offers to help are
not answered in a timely way. For example, Mailpile's visibility
might be increased if it were submitted in its present state to
Debian Unstable. Within Debian, Mailpile might be a good
addition to the [FreedomBox](https://en.m.wikipedia.org/wiki/FreedomBox) project.

I'm posting this as a Github issue, not a Community post,
because most of the possible volunteers are accessible through
Github handles. I've tried to mention most of the recent
contributors in the following list so they will get the post. If
I have missed anyone, my apologies.

@aviau
@AlexanderHaase 
@JocelynDelalande
@smari
@peturingi
@loidor
@infinite-Joy
@grekko
@noahhaasis
@FlakyBit
@largestprime
@mgax
@kolibrik7
@futal
@dorumd
@comradekingu
@rgarrigue
@kenjones21
@hiteshramani
@balping
@MikeynJerry
@Epse
@DaLynX
@nkfarwell
@deepak21188
@balki
@philratcliffe
@lee7670
@karnation22
@mgax
@yegortimoshenko
@vezult
@nik1082
@wiktor-k
@foursixnine
@JackDca

**Comment by yegortimoshenko** [comment]
I personally would like to participate, with the most important short-term goals (for me personally) of porting Mailpile to Python 3 (see #2150) and implementing session key storage (see #2155). However, so far I don't have much any contributions that have already landed in.

**Comment by BjarniRunar** [comment]
Thanks for filing this @JackDca. And thanks (in advance) to anyone who would like to step up.

Just as a reminder, the IRC channel is still open (Freenode #mailpile) and I'm happy to discuss things informally there if I'm at the computer. The other venue is our Discourse forum, probably the Development category: https://community.mailpile.is/c/development - that would be the most appropriate place for this, really, but this is still a fine start. :heart:

**Comment by noahhaasis** [comment]
Since I don’t have a deep understanding of the Codename yet I would like to start in the first category and maybe transition later when I’m more familiar with the Codename and the project.

**Comment by aviau** [comment]
I just wanted to leave a small word of encouragement and state that I am coincidentally wearing my Mailpile shirt today <3!

![image](https://user-images.githubusercontent.com/2706882/60983731-7cc71b00-a308-11e9-9b52-b90d75b92870.png)

On the subject of this issue, I intend to keep helping with packaging tasks if needed. I have been ready to upload Mailpile to official Debian repositories for a while but I felt like the project was not ready to be in the archive yet. You can definitely count me in for 1.0.

**Comment by Kabouik** [comment]
Please add me to the list. I'm in no way capable of coding anything, this is not my area of expertise, and unfortunately I don't have much time to invest in the project either because I'm not far from being burnt out with my own occupation. However, I care deeply about Mailpile too, have been using it for years, and would be happy to follow and participate in discussions, however minor my input might be. At least I'm curious and want to follow what is happening. This is also why I'm subscribed to all Mailpile issues (although I have zillions of unread email alerts for github issues).

**Comment by h3artbl33d** [comment]
Count me in, **definitely**. Though I am experienced with Python development, I'd like to start with support and _auditing_ the code. I want to feel familiar with the code before starting to hack on it ;) If I had to pick, I'd like to hack on the core once I am confident enough with the existing codebase.

- **I am able and willing to contribute:** time and effort, code, documentation, support and localization.
- **Aspects I explicitly pass on:** packaging and containerization/docker/lxc.

**Comment by comradekingu** [comment]
Would still like to work on the language of Mailpile. I think the preferable solution is to have two different people working at it, and asking Bjarni to explain things if needed.
It would help greatly in that regard to move the translation effort onto https://weblate.org/hosting then I could help out there too, and even administrate things.

**Comment by h3artbl33d** [comment]
> It would help greatly in that regard to move the translation effort onto https://weblate.org/hosting then I could help out there too, and even administrate things.

Just curious, what would Weblate add that Transifex doesn't have?

**Comment by BjarniRunar** [comment]
@comradekingu So as not to derail this thread, can you file a new issue to discuss your translation ideas? It would be absolutely great to have an i18n-manager, and I think you're experienced enough with the app and the tech that you could do a great job.

**Comment by JackDca** [comment]
Thank you for all the responses, everyone.

I am travelling today and will catch up tomorrow.

**Comment by BjarniRunar** [comment]
Just a quick reminder, folks! This project has a code of conduct, please be sure you are comfortable with it before investing too heavily here: https://github.com/mailpile/Mailpile/blob/master/CODE_OF_CONDUCT.md

(I know this matters to some people, so this is just to clarify that the CoC is not up for debate at this point in time. Please respect that. There are lots of projects out there, agree to disagree, move along...)

**Comment by JackDca** [comment]
Thanks again for the responses!

Mailpile uses Python, Javascript, HTML, CSS, PGP, i18n among other technologies. Also it's targeted at Linux, Windows, and macOS. Another need is for user and developer documentation - Docs. So it would help to hear from each person, what is your level of knowledge/experience in each of these areas - or other major areas if yours aren't listed?

For me, if 0 is none and 5 is a lot:
Python - 2
Javascript - 1
HTML - 1
CSS - 1
PGP - 2
i18n - 0
Docs - 4
Linux - 4
Windows - 4
macOS - 0
also C/C++ - 5

Here's what I've seen so far:
@yegortimoshenko - Python
@noahhaasis - Tell us!
@aviau - Linux packaging
@Kabouik - Python
@h3artbl33d - Python
@comradekingu - i18n

**Comment by h3artbl33d** [comment]
Could you clarify some points? For example, how is the scoring on the OS aspects interpreted (from which viewpoint). And with i18n, are you referring to the implementation of translation in the core, translating Mailpile into other languages, etc?

I have extensive knowledge of BSD, Linux and macOS, I'd rate these a 5. In Windows I definitely know my way around, but I fail miserably with PowerShell - I'd give myself a 2. I can code in Python (4 or 5), script in Bash (5) and build some fancy things with HTML and CSS (4). I am at ease with PGP/GPG (3), but not with implementing it properly in code. I value good documentation (5) and am able to write it as well.

**Comment by JackDca** [comment]
@h3artbl33d - at this point I was hoping to encourage those who were interested in participating to informally indicate their areas of expertise. I hadn't defined any precise scale. Your summary of your own capabilities more than answers my question. Thanks for your response!

**Comment by vezult** [comment]
On Wed, Jul 10, 2019 at 08:00:42AM -0700, Jack Dodds wrote:
> In @BjarniRunar's blog post [Burned Out and Happy?](https://www.mailpile.is/blog/2019-04-06_Burnout.html) he says,
>  "I care deeply about this project, but it really is too big for one
> person and I haven't successfully built a team to support me."
> 
> I too have spent some time on Mailpile (but nothing like
> Bjarni's contribution!). I believe that it is an important
> effort and want to see it continue. Can this community have a
> discussion about action we can take to create the team as
> suggested by Bjarni?
> 
> In the world of engineering management, it is recognized that
> the most effective way to get a project done is often to have a
> single person do everything from function definition to detailed
> design and testing. No time is used up in communicating between
> team members, debating interfaces and so on. But this only works
> to the point where the project is small enough for one person.
> Although Mailpile has had many contributors, Bjarni's share far
> exceeds anyone else's so it is not far from being a one person
> project, but it has grown past the point where that structure
> works.
>  
> Creating a workable project team is not a trivial task, but it
> seems desirable at this point. Here are some ideas. I'm sure
> that they are not new, but it would be good to see some
> discussion.
> 
> 1. Most important, will any of us make an ongoing commitment
> to write or modify code to fix specific issues? It might help if
> issues could be prioritized and contributors recruited for e.g.
> the five highest priority issues.

I've done some work toward using firejail to sandbox calls to external
utilities. Events in RL conspired to derail my work on that, and I
never got back to it. I'd like to finish that work.

I don't have a lot of time to devote, so I don't know that I can
pledge to contribute in a sustained/consistent way.

-davidc

**Comment by h3artbl33d** [comment]
> I've done some work toward using firejail to sandbox calls to external utilities. Events in RL conspired to derail my work on that, and I never got back to it. I'd like to finish that work. I don't have a lot of time to devote, so I don't know that I can pledge to contribute in a sustained/consistent way.

That is actually rather interesting and neat. I was already tinkering with writing some docs on paranoia setups. Besides sandboxing calls to external utilities, users can jail/sandbox/pledge/unveil their browsers - which pose a rather large attack vector. The browser being unable to touch the Mailpile data folder is one extra defense barrier.

Would you be open to discuss this further and see whether we can contribute together?

**Comment by spiraldancing** [comment]
Hello all,

Just a quick intro ... I'm a (mostly-)Microsoft developer with 15-ish years under my belt, primary experience is in C# / ASP.NET / T-SQL, with a healthy dose of HTML/CSS/js, and a smattering of 4 dozen other things (I've worn many hats).

The past couple of years, I've been trying to pivot away from the M$ world (mainly because I have no respect for the company), and into Linux.  I'm a Big Fan of FOSS, and for many years, I've been wanting to start diving into FOSS development, help out projects I respect (and, generally, use).  I've made some minor contributions along the way, but never really contributed the way I feel I could/should.

I've started feeling my way around here at Mailpile, looking at the project and the community, and I'd like to help.  Still not sure how best to do that, but I'll either figure it out soon enough, or at least figure out what questions to ask (and ideally, who to ask them at).

Caveat ... I tend to get excited with new projects, jump in enthusiastically for a little while, then get distracted by Life and fade away from said projects.  So, maybe don't put me in charge of anything mission-critical for the first few weeks ... just in case.  ;-P

**Comment by foursixnine** [comment]
Hi! I finally managed to squeeze in some time to answer. I'm mainly a Perl Developer/QA Engineer, but I know my ways around code (more or less). So if there's some direction on where to take the project, I'd be more than happy to contribute (And finally switch fully to Mailpile :D)

Regarding skills and stuff, idk, I think I can do stuff at best, worst case scenario I can always set up some functional tests or stuff like that, when needed. Thanks for all the effort @BjarniRunar and thanks @JackDca for stepping in!

**Comment by JackDca** [comment]
Thanks for responding @foursixnine. Good to see QA experience! Someone who contributes code must thoroughly debug and test it before committing, but it's important to have another level of testing before the code is merged into master. I'm working on suggesting ways that we can use all the skills that have been identified.

**Comment by JackDca** [comment]
Hello @spiraldancing - thanks for your response!
I agree that the best way to start is to install and use Mailpile for your own email. I'm curious - what kind of platform would you be using? I don't have any data to support this, but my guess is that most people in the community are using some flavour of Gnu/Linux. If you are using Windows for your personal computing (including Mailpile), that might add some useful diversity to the testing of Mailpile.
I have Win7 and Win10 VMs on my Debian box and I have run Mailpile on both. But, MP was set up on those systems using a old packaging system that ran on Linux (now superseded by packages/windows-wix packing system which I have not used). I would answer any questions to the best of my ability.

**Comment by JackDca** [comment]
**Update:**

In response to this issue, one group has agreed to work on Issue #2180, which is on the list of things to be done before before the 1.0 release. More details can be found at the [Mailpile Discourse topic](https://community.mailpile.is/t/work-on-issue-2180-saved-search-ui-is-broken/227) that has been created to track the work of that group.

There are several other issues on the [1.0 blocker list](https://github.com/mailpile/Mailpile/milestone/4) that have not been spoken for. Any one of these would be a good way to contribute to the Mailpile development. The biggest need is for people to write and patch the code, but reviewers and testers are also vital. Anyone who is interested could comment here, or at the [Mailpile Discourse site](https://community.mailpile.is/c/development) or on IRC at irc://irc.freenode.net/#mailpile.

I'm mentioning the following contributors so that they will see this update:

@loidor   @infinite-Joy   @grekko   @FlakyBit   @largestprime   @mgax   @kolibrik7   @futal   @dorumd   @rgarrigue   @kenjones21   @hiteshramani   @balping   @MikeynJerry   @Epse   @DaLynX   @nkfarwell   @deepak21188   @balki   @philratcliffe   @lee7670   @karnation22   @mgax   @nik1082   @wiktor-k

**Comment by spiraldancing** [comment]
Hello @JackDca 

First off, I got side-tracked by Life again.  Apologies for the slow response.  (Although, oddly, the GitHub email notification just prompted me now, in response to your latest "Update:" post, and not the post where you actually mentioned me ... ?)

Regardless, I'm back now and paying attention again.  Despite my Microsoft professional background, I too, run various flavors of Linux at home, almost exclusively.  My primary machine has a Win07 partition on it, mainly for Skyrim gaming, and I have a couple of Win07 VMs for whatever M$ dev work that comes up.

My primary task (starting today) will be to just walk thru the MP installation process.  I'll be looking at the (user) documentation along the way, and making notes on any issues there (in my experience, accurate directions for non-techie users is almost always the least-supported part of FOSS projects).  I can certainly try installing it on a Win07 box, as well.

I'll probably need several days (if not weeks) just to get familiar with the code base, and the basic FOSS distributed dev process and such-like ... so it'll be a little while before I'm ready to do real hands-on bug-fixes.  So I'm definitely not claiming it yet, but tentatively, I may take a look at [#2152](https://github.com/mailpile/Mailpile/issues/2152), since I run a couple of Debian VMs, plus mostly use Debian-downstream distros ... unless anyone wants to recommend a better "first effort" issue?

BTW & PS:  I'm not seeing any kind of prioritization methodology ... are all "1.0 Release" open issues effectively equal priority?  Or is there some kind of weighting system I'm overlooking?

Thanks,
Eric

**Comment by JackDca** [comment]
@spiraldancing - thanks for your response. We are all volunteers, no apology is necessary!

**Priorities:** One of the rewards of being a volunteer is that you can choose your work. As I see it, reaching the 1.0 release would be a big boost for everyone involved, so I'm choosing issues I'd like to work on, but mainly from the 1.0 blockers. I don't see a big priority difference between them (and it's not for me to set priorities anyway).

**Notes:** Agreed that directions and documentation are a useful contribution! When I learn something about the Mailpile internals I try to document it in the wiki - [example](https://github.com/mailpile/Mailpile/wiki/Data-structures%3A-Metadata-index). Also, for your notes, you could create a topic in the [Development category of the Discourse forum](https://community.mailpile.is/c/development) as suggested by @BjarniRunar. This Issue #2223 can remain a meta-issue about facilitating the creation of a team.

**Schedules:** One of the disagreeable experiences of commercial engineering development is getting "behind schedule". It happens because the time required to to do development tasks is usually - no, **always** - underestimated. Since we are volunteers, I say we should make the experience more rewarding by eliminating schedules! "Several days" or several weeks - you are making a contribution, that is what matters. (Please excuse the rant.)

**Getting familiar with the code:** In my experience this takes a lot of time. Reading code does not work for me. Instead I like to focus on some specific function, run Mailpile in a console, and add debug print:
`    print 'AAAAB', self._loop_count     # *** DEBUG`

or to sort out what calls what:
`        print 'AAAAC'`
`    traceback.print_stack()             # *** DEBUG`

Note: if the module has been "futurized" as indicated by `from __future__ import print_function` at the top, then `print` must be in functional form, e.g.
`    print('CCCCC', time.time())      # *** DEBUG`

I use [Eclipse](https://packages.debian.org/stretch/eclipse) as a super-editor which lets me do global searches (to find my way around), source code comparisons (with unmodified code, git branches and tags) and quickly revert my debug code when necessary.

Hope that helps!

**Comment by spiraldancing** [comment]
Thanks @JackDca 

All good info.  And yeah, if the codebase is as large as it sounds, I may never be familiar with all of it ... goal is to just get the overall lay of the land, the build/deploy process, and then just really familiarize myself with the source code in the area(s) I actually dive into.

"futurized"?  Meaning, fully converted to Python 3.x?

For most of my development, I'm just starting to use VSCode (actually, the more FOSS-y fork called VSCodium).  I'm still learning the IDE's ins and outs.  Any reason it's not a good tool to work on this project?  Or any other general 'gotchas' to watch out for?

Danke, y gracias,
e

**Comment by JackDca** [comment]
@spiraldancing - see my response on the Mailpile Discourse site - [Setting up tools for Mailpile development](https://community.mailpile.is/t/setting-up-tools-for-mailpile-development/233).

**Comment by reyman** [comment]
As Mailpile is the only self-hosted web client that support GPG encryption with fast searching, i don't understand why there are so few sponsor ? :-/ Is there some plan to relaunch a kick/indiegogo campaign ? Devs. need to be paid for the works, especially i suppose for huge and complex project like Mailpile ... Other FOSS project take the path of Patreon/Git Sponsoring/Donation/etc ( git-annex, photoprism, etc ?) to provide support and quality after release, could we imagine something like that with Mailpile ? But first, we need some "news" / "communication" to say to the world "hey this project works" (like this great unoficial tutorial for newcomers (like me) on discourse  : https://community.mailpile.is/t/mailpile-tutorial-for-newcomers/597 ), "we are not dead", "we need your support" !
